cli/dcoscli/package/main.py

```diff
@@ -206,6 +215,235 @@ def _remove_repo(repo_name):
+
+    build_definition_directory = os.path.dirname(build_definition_path)
+
+    # get the path to the output directory
+    if output_directory is None:
+        output_directory = os.path.join(build_definition_directory, "target")
+
+        # ensure the directory exists
+        os.makedirs(output_directory, exist_ok=True)
+
+    logger.debug("Using [%s] as output directory", output_directory)
+
+    if not os.path.exists(build_definition_path):
+        raise DCOSException(
+            "The file [{}] does not exist".format(build_definition_path))
+
+    # load raw build definition
+    with util.open_file(build_definition_path) as bd:
+        build_definition_raw = util.load_json(bd, keep_order=True)
+
+    # validate DC/OS Package Build Definition with local references
+    build_definition_schema_path = "data/schemas/build-definition-schema.json"
+    build_definition_schema = util.load_jsons(
+        pkg_resources.resource_string(
+            "dcoscli", build_definition_schema_path).decode())
+
+    errs = util.validate_json(build_definition_raw, build_definition_schema)
+
+    if errs:
+        raise DCOSException('Error validating package: '
+                            '[{}] does not conform to the specified '
+                            'schema'.format(build_definition_path))
+
+    # resolve local references in build definition
+    _resolve_local_references(
+        build_definition_raw,
+        build_definition_schema,
+        build_definition_directory
+    )
+
+    # at this point all the local references have been resolved
+    build_definition_resolved = build_definition_raw
+
+    # validate resolved build definition
+    package_schema_path = "data/schemas/package-schema.json"
+    package_schema = util.load_jsons(
+        pkg_resources.resource_string(
+            "dcoscli", package_schema_path).decode())
+
+    errs = util.validate_json(build_definition_resolved, package_schema)
+
+    if errs:
+        raise DCOSException('Error validating package: '
+                            'one of the references in [{}] does '
+                            'not conform to the specified '
+                            'schema'.format(build_definition_path))
+
+    # create the manifest
+    manifest_json = {'built-by': formatted_cli_version()}
+
+    # create the metadata
+    metadata_json = build_definition_resolved
+
+    # create zip file
+    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+        tmp_name = temp_file.name
+        with zipfile.ZipFile(
+                temp_file.file,
+                mode='w',
+                compression=zipfile.ZIP_DEFLATED,
+                allowZip64=True) as zip_file:
+
+            metadata = json.dumps(metadata_json, indent=2).encode()
+            zip_file.writestr("metadata.json", metadata)
+
+            manifest = json.dumps(manifest_json, indent=2).encode()
+            zip_file.writestr("manifest.json", manifest)
+
+    try:
```

Comment:
Can you explain why we are doing this outside of the `ContextManager` for `temp_file`?